File: src/bin/shred_probe.rs
//! Standalone ShredStream feed probe — run on the co-located box to confirm
//! the fast feed is live and hitting our pools before wiring it into the
//! shadow harness. `RUN_MS=60000 cargo run --release --bin shred_probe`.

use std::time::Duration;
use tokio::sync::mpsc;

#[tokio::main]
async fn main() {
    let _ = dotenvy::dotenv();
    let port: u16 = std::env::var("SHREDSTREAM_PORT")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(20000);
    let run_ms: u64 = std::env::var("RUN_MS")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(60_000);
    println!("shred-probe — listening udp/{port} for {}s…\n", run_ms / 1000);

    let (tx, mut rx) = mpsc::unbounded_channel();
    let _handle = arb_engine::shredstream::run_shredstream_feed(port, tx);

    let mut count: u64 = 0;
    let deadline = tokio::time::sleep(Duration::from_millis(run_ms));
    tokio::pin!(deadline);
    loop {
        tokio::select! {
            _ = &mut deadline => break,
            Some(t) = rx.recv() => {
                count += 1;
                if count <= 20 || count % 100 == 0 {
                    let sig = &t.sig[..t.sig.len().min(8)];
                    println!("trigger #{count} {} slot {} sig {sig}…", t.venue, t.slot);
                }
            }
        }
    }
    println!("\nshred-probe: {count} pool triggers in {}s", run_ms / 1000);
}

File: src/bin/shredProbe.ts
// Port of src/bin/shred_probe.rs
//
// Standalone ShredStream feed probe — run on the co-located box to confirm
// the fast feed is live and hitting our pools before wiring it into the
// shadow harness. `RUN_MS=60000 tsx src/bin/shredProbe.ts`.

import 'dotenv/config';
import { runShredstreamFeed, type Trigger } from '../lib/shredstream.js';

async function main(): Promise<void> {
  const port = Number.parseInt(process.env.SHREDSTREAM_PORT ?? '', 10) || 20000;
  const runMs = Number.parseInt(process.env.RUN_MS ?? '', 10) || 60_000;
  // ALT resolution needs an RPC; set RPC_ENDPOINT to catch routed swaps.
  const rpc = process.env.RPC_ENDPOINT;
  console.log(`shred-probe — listening udp/${port} for ${runMs / 1000}s (ALT resolution: ${rpc !== undefined ? 'on' : 'off'})…\n`);

  let count = 0n;
  const triggers: Trigger[] = [];
  let wake: (() => void) | undefined;
  runShredstreamFeed(port, rpc, (t) => {
    triggers.push(t);
    if (wake) {
      const w = wake;
      wake = undefined;
      w();
    }
  });

  const deadline = Date.now() + runMs;
  await new Promise<void>((resolve) => {
    const tick = (): void => {
      const now = Date.now();
      if (now >= deadline) {
        resolve();
        return;
      }
      if (triggers.length > 0) {
        const t = triggers.shift()!;
        count += 1n;
        if (count <= 20n || count % 100n === 0n) {
          const sig = t.sig.slice(0, Math.min(8, t.sig.length));
          console.log(`trigger #${count} ${t.venue} slot ${t.slot} sig ${sig}…`);
        }
        tick();
        return;
      }
      const timer = setTimeout(tick, Math.min(50, deadline - now));
      wake = () => {
        clearTimeout(timer);
        tick();
      };
    };
    tick();
  });

  console.log(`\nshred-probe: ${count} pool triggers in ${runMs / 1000}s`);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
